src/commands/SinceCommand.ts

```diff
@@ -126,101 +74,78 @@ function formatResult(action: string, targetRoomId: string, recentJoins: Join[],
 // - resolves any room alias into a room id;
 // - attempts to execute action;
 // - in case of success, returns `{ok: undefined}`, in case of error, returns `{error: string}`.
-async function execSinceCommandAux(destinationRoomId: string, event: any, mjolnir: Mjolnir, tokens: ParseEntry[]): Promise<Result<undefined>> {
-    const [dateOrDurationToken, actionToken, maxEntriesToken, ...optionalTokens] = tokens;
-
-    // Parse origin date or duration.
-    const minDateResult = parseToken("<date>/<duration>", dateOrDurationToken, source => {
-        // Attempt to parse `<date>/<duration>` as a date.
-        let maybeMinDate = new Date(source);
-        let maybeMaxAgeMS = Date.now() - maybeMinDate.getTime() as number;
-        if (!Number.isNaN(maybeMaxAgeMS)) {
-            return { ok: { minDate: maybeMinDate, maxAgeMS: maybeMaxAgeMS} };
-        }
-
-        //...or as a duration
-        maybeMaxAgeMS = parseDuration(source);
-        if (maybeMaxAgeMS && !Number.isNaN(maybeMaxAgeMS)) {
-            maybeMaxAgeMS = Math.abs(maybeMaxAgeMS);
-            return { ok: { minDate: new Date(Date.now() - maybeMaxAgeMS), maxAgeMS: maybeMaxAgeMS } }
-        }
-        return { error: "" };
-    });
-    if ("error" in minDateResult) {
-        return minDateResult;
-    }
-    const { minDate, maxAgeMS } = minDateResult.ok!;
-
-    // Parse max entries.
-    const maxEntriesResult = parseToken("<maxEntries>", maxEntriesToken, source => {
-        const maybeMaxEntries = Number.parseInt(source, 10);
-        if (Number.isNaN(maybeMaxEntries)) {
-            return { error: "Not a number" };
-        } else {
-            return { ok: maybeMaxEntries };
-        }
-    });
-    if ("error" in maxEntriesResult) {
-        return maxEntriesResult;
+async function execSinceCommandAux(destinationRoomId: string, event: any, mjolnir: Mjolnir, lexer: Lexer): Promise<Result<undefined>> {
```

Comment:
rather than passing the lexer to commands, wouldn't it be better if commands declared what arguments they are expecting (can be part of the same api to register with the handler)? This means that they don't need to know implementation detail of the lexer

Reply:
I would like to move towards what you suggest. I'm not sure I'll succeed in a single pass, though.